### DISCUSSION TITLE
Systemd after security hardening, how to connect to redis sock file

### DISCUSSION THREAD

**Discussion by khangp0000** [authelia/authelia#6972] [Q&A]
After the CR for systemd sandbox hardening [(link)](https://github.com/authelia/authelia/commit/648fdfe93b5ac24f49624c4c24fc1cee4bce398b), I need to add following to systemd override to allow  Authelia service to read redis socket file:
```
[Service]
PrivateUsers=no
CapabilityBoundingSet=~
```

Is there a better way to do this? 
- For `PrivateUsers`, the  redis socket file is owned by `redis:redis`, which when execute under `PrivateUsers=yes` it turn into `nobody:nobody` which authelia does not have permission to access.
- For `CapabilityBoundingSet`, I tried adding different capability but cannot find which one it needs to works, keep getting `dial unix /var/run/redis/redis.sock: connect: permission denied`

**Comment by glime-ai** [comment]
Please note this response was created using generative AI leveraging previously available information from the repository. This information may be helpful however it is our strong recommendation that you carefully consider the information before blindly applying it.

---

To address the issues you're encountering with systemd and Authelia's access to the Redis socket file after the systemd sandbox hardening, let's consider each point:

1. **PrivateUsers**: This setting isolates the service from the system's users and groups. When `PrivateUsers=yes` is set, the service has its own user namespace, making it unable to see or interact with users and groups defined on the host system. This is why the Redis socket file, owned by `redis:redis` on the host, appears to be owned by `nobody:nobody` within the isolated namespace. Disabling `PrivateUsers` (setting it to `no`) is a straightforward way to allow Authelia to access the Redis socket file by seeing the host's user and group IDs. An alternative approach, although more complex and might not be directly supported by Authelia or Redis configurations, involves mapping the `redis` user and group into the service's namespace, which systemd currently has limited support for.

2. **CapabilityBoundingSet**: This directive controls which capabilities are available to the service. The error `dial unix /var/run/redis/redis.sock: connect: permission denied` suggests a lack of necessary permissions to access the socket file, but it's more related to file system permissions rather than missing capabilities. Capabilities are more about granting specific privileges that are not directly related to file access permissions.

Given the context and the specific issues you're facing, your approach to modifying the systemd service file by setting `PrivateUsers=no` seems to be the most practical solution to ensure Authelia can access the Redis socket file. As for `CapabilityBoundingSet`, since the problem is more about file system permissions and user/group visibility due to namespaces, adjusting capabilities might not resolve the access issue.

If maintaining the security benefits of `PrivateUsers=yes` is crucial, you might consider using TCP/IP communication with Redis instead of Unix sockets, if your security model allows it. This would circumvent the need to access the file system for socket communication, although it might introduce other considerations for securing the TCP/IP connection.

Unfortunately, without more advanced user namespace mapping features in systemd or changes in how Authelia and Redis handle user and group IDs in these scenarios, the options for resolving this are somewhat limited to the workaround you've identified.

This comment was generated by [Glime](https://glimelab.ai/?utm_source=6fsHkcyufjQLaB&utm_medium=github).

**Comment by lukasmalkmus** [comment]
Facing the same issue. Thanks for the tip @khangp0000, this worked for me, so far!

**Comment by e64462** [comment]
same issue, OP's proposed workaround also resolves the issue for me.

I also tested the AI solution by using 
`PrivateUsers=yes` (leaving `CapabilityBoundingSet` unchanged)
`CapabilityBoundingSet=~` (leaving `PrivateUsers` unchanged)

neither resolves the issue though. I believe the second test means that no combination of capabilities could be granted with `PrivateUsers=yes` to resolve the issue. 

Apparently using `CapabilityBoundingSet` automatically enables `PrivateUsers`
source: https://www.freedesktop.org/software/systemd/man/latest/systemd.exec.html#Capabilities
> These options are only available for system services, or for services running in per-user instances of the service manager in which case PrivateUsers= is implicitly enabled
